src/librbd/cache/pwl/AbstractWriteLog.cc

```diff
@@ -1342,6 +1347,10 @@ void AbstractWriteLog<I>::complete_op_log_entries(GenericLogOperations &&ops,
                       log_entry->ram_entry.write_bytes);
     m_perfcounter->tinc(l_librbd_pwl_log_op_app_to_cmp_t, now - op->log_append_start_time);
   }
+  if (need_update_state) {
```

Comment:
@idryomov I have also thought of using `unique_lock` instead of `lock_guard` to facilitate free locking and unlocking.  However, there are several places where the `lock_guard` is not easy to change, such as here. I saw that you still keep the original `lock_guard` here, and then lock `unique_lock` when `lock_guard` is unlocked. This is different from the previous behavior of keeping `unique_lock` in the whole process. But if this brief release and re locking is allowed, then this may be a better approach.

Reply:
> This is different from the previous behavior of keeping `unique_lock` in the whole process. But if this brief release and re locking is allowed, then this may be a better approach.

Correct, but I derived it from your change and the surrounding code.  I concluded that:
- releasing `m_lock` is OK based on your proposed change (this part was present in the previous revision of this PR)
- re-acquiring `m_lock` is OK based on the fact that it is re-acquired for calling `wake_up` anyway (this part was present in existing code before this PR)

Honestly, it feels like `AbstractWriteLog` locking is really messy but I don't think we would be making it worse with this change.  Do you see any holes in this argument?

Reply:
Agree. Locks in PWL are messy.  Thanks your effort.